libraries/lib-numeric-formats/formatters/BeatsNumericConverterFormatter.cpp

```diff
@@ -0,0 +1,235 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/**********************************************************************
+
+ Audacity: A Digital Audio Editor
+
+ @file BeatsNumericConverterFormatter.cpp
+
+ Dmitry Vedenko
+
+ **********************************************************************/
+#include "BeatsNumericConverterFormatter.h"
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+
+#include "NumericConverterRegistry.h"
+
+#include "SampleCount.h"
+
+namespace
+{
+
+class BeatsFormatter final :
+    public NumericConverterFormatter
+{
+public:
+   BeatsFormatter(
+      double tempo, int upperTimeSignature, int lowerTimeSignature,
+      int fracPart)
+       : mTempo { tempo }
+       , mUpperTimeSignature { upperTimeSignature }
+       , mLowerTimeSignature { lowerTimeSignature }
+       , mFracPart { fracPart }
+   {
+      mFields.reserve(mFracPart > 0 ? 3 : 2);
+
+      // Range is assumed to allow 999 bars.
+      auto& barsField = mFields.emplace_back(NumericField { 1000, true });
+      barsField.label = L" " + XO("bar").Translation() + L" ";
+
+      // Beats format is 1 based. For the time point "0" the expected output is "1 bar 1 beat [1]"
+      // For this reason we use (uts + 1) as the "range".
+      // On top of that, we want at least two digits to be shown. NumericField accepts range as in
+      // [0, range), so add 1.
+
+      auto& beatsField = mFields.emplace_back(
+         NumericField { std::max<size_t>(12, mUpperTimeSignature + 1), true });
+      beatsField.label = L" " + XO("beat").Translation();
+
+      if (mFracPart > 0)
```

Comment:
I'm not convinced yet that there is no path that divides by `mFracPart` when it is zero.  There is a test here, but you don't then reassign it as 1 or anything like that.

Reply:
There is no integer division. There cannot be any exception. In any case, please do not continue the review now, specifically of this file.

Reply:
> There is no integer division. There cannot be any exception. In any case, please do not continue the review now, specifically of this file.

Yes, `const auto fracLength = beatLength * mLowerTimeSignature / mFracPart;` is not an integer division because `mFracPart` is coerced to floating point.

But that is irrelevant and  my objection is still unanswered.  I do not see a guard against division by `mFracPart` when it is zero.

Reply:
It. Is. Safe.

And again - do not review this file now.